### PR TITLE
storage: campaign idle replicas eagerly

### LINKED ISSUE #9273: stability: slow performance after restarting cluster with many ranges
When we restart gamma and beta, performance is bad for ~5-10 minutes before recovering. I suspect this is due to lazy raft loading. Both clusters have 8.39k ranges. Due to lazy raft loading we won't initialize the raft group for a range until it is accessed. `block_writer` touches ranges randomly and we currently have it configured to perform 50 concurrent requests on these clusters. I'm not sure if I'm following the Raft initialization, but it looks like every range will have to wait for an election because every replica starts as a follower. With 8.39k ranges and 50 concurrent ops we'd expect 167 rounds of waiting for the election timeout and then campaigning. With our current 200ms tick interval and 15 tick election timeout, this translates into 3 sec per round and ~8.5 min for all of the ranges to be woken up.

I'm not sure if this is worth addressing in the short term.

Cc @cockroachdb/stability

### PR DESCRIPTION
**Pull request by petermattis**
Fixes #9273.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9276)

### CODE REVIEW

**Comment by bdarnell** [comment]
This is only a partial solution for #9273, but that's OK because I don't think #9273 is that important (cold restarts will mostly go away after beta). This is more useful when you consider the general case of quiescent ranges (especially when we let them go back to sleep); we're going to need to be able to reconstitute raft groups relatively quickly. 

It's a partial solution because requests that arrive during the first three-second window will have to wait out the entire election timeout starting at that point, but requests arriving just after the three-second window will be able to kick off an election (waking up the earlier requests in the process). 

I think this is worth merging if the start of the timer is moved somewhere appropriate. 

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1652 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRRNJqnAWfNPFzAggIV:-KRRNJqnAWfNPFzAggIW:blvd87y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bf497ce9e2b339be4cddb34f7b3ae08b153f28/storage/replica.go#L1652)):*
> ```Go
> 	//
> 	// TODO(peter): Perhaps that eager campaign logic should be unified with the
> 	// logic here.
> ```

Yeah, I think we could put this in `withRaftGroupLocked` if we add the condition that `SoftState.Lead == 0`. 

---

*[storage/store.go, line 1142 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRRIs0pHfu-Lq5i2CKe:-KRRIs0qaGOfzBsbC2Sc:baog2vm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bf497ce9e2b339be4cddb34f7b3ae08b153f28/storage/store.go#L1142)):*
> ```Go
> 		return err
> 	}
> 	// Once we have gossiped the store descriptor the first time, other nodes
> ```

It's actually the node descriptor, not the store descriptor, that matters for this.

We can know that we've sent something into gossip, but not that other nodes have received it. We start gossiping store descriptors before we have verified that we can receive anything from the gossip network. I think a better place to start the timer is in `server/Node.connectGossip`, or maybe the later of `connectGossip` and the first gossip of the node descriptor (I think they happen in that order, but the code is a little hard to follow and it's unclear what's guaranteed vs incidental). 

---

*[storage/store.go, line 1163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRRKjEUdvIHNVGaKvBX:-KRRKjEUdvIHNVGaKvBY:b-4bothu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bf497ce9e2b339be4cddb34f7b3ae08b153f28/storage/store.go#L1163)):*
> ```Go
> 		return false
> 	}
> 	return !timeutil.Now().Before(s.idleReplicaElectionTime.at)
> ```

This loses the randomized element of the real raft election timeout (it picks a value between `RaftElectionTimeoutTicks` and `2*RaftElectionTimeoutTIcks`), which means that these elections are somewhat more likely to become contested (assuming traffic is distributed evenly across a cluster that is restarted simultaneously). But I think that's OK; it just adds a network round trip or two to the process since a contested election just restarts the clock to where it would have been anyway if we weren't doing this campaigning. (and it gets even better if we implement the "separate follower and candidate timeouts" optimization from the Raft Refloated paper).

This could have an effect of concentrating leadership (e.g. of the stats ranges) on the first node to wake up, but our leases already do that anyway. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-:-KRRN_WHJjoRj_Yozn9a:b-nnhfb2)*

**Comment by petermattis** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1652 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRRNJqnAWfNPFzAggIV:-KRT_mb9D-V8oKjKtWJn:b-urzo6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bf497ce9e2b339be4cddb34f7b3ae08b153f28/storage/replica.go#L1652)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Yeah, I think we could put this in `withRaftGroupLocked` if we add the condition that `SoftState.Lead == 0`. 
</details>
Just tried to do that and failed. Putting this campaigning in `withRaftGroupLocked` caused every replica to campaign. The reason I put the campaigning here in the first place is that I only want the proposer to campaign. Perhaps I just did it wrong. 

---

*[storage/store.go, line 1142 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRRIs0pHfu-Lq5i2CKe:-KRTxlITi4ma0BncrgAc:bq964fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bf497ce9e2b339be4cddb34f7b3ae08b153f28/storage/store.go#L1142)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It's actually the node descriptor, not the store descriptor, that matters for this.
> 
> We can know that we've sent something into gossip, but not that other nodes have received it. We start gossiping store descriptors before we have verified that we can receive anything from the gossip network. I think a better place to start the timer is in `server/Node.connectGossip`, or maybe the later of `connectGossip` and the first gossip of the node descriptor (I think they happen in that order, but the code is a little hard to follow and it's unclear what's guaranteed vs incidental). 
</details>
`Store.GossipStore` is called after both `Node.connectGossip` and the gossiping of the node descriptor (which usually happens via `Node.validateStores -> Node.initNodeID -> Gossip.SetNodeDescriptor`). I could see about moving computing the timeout to earlier, but performing it here is both conservative and somewhat convenient.

---

*[storage/store.go, line 1163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRRKjEUdvIHNVGaKvBX:-KRTyBiIli3vUc7e_jvW:bicimqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bf497ce9e2b339be4cddb34f7b3ae08b153f28/storage/store.go#L1163)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This loses the randomized element of the real raft election timeout (it picks a value between `RaftElectionTimeoutTicks` and `2*RaftElectionTimeoutTIcks`), which means that these elections are somewhat more likely to become contested (assuming traffic is distributed evenly across a cluster that is restarted simultaneously). But I think that's OK; it just adds a network round trip or two to the process since a contested election just restarts the clock to where it would have been anyway if we weren't doing this campaigning. (and it gets even better if we implement the "separate follower and candidate timeouts" optimization from the Raft Refloated paper).
> 
> This could have an effect of concentrating leadership (e.g. of the stats ranges) on the first node to wake up, but our leases already do that anyway. 
</details>
Should I make the timeout `2*RaftElectionTimeoutTicks`? That would mean there is a 6 second window at startup where we'll have to wait for campaigning to occur, but that seems ok. Or leave as is and add a comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276)*

**Comment by petermattis** [comment]
Reworked the code a bit and added a test of `Store.canCampaignIdleReplica`. Not sure how to provide a system-level test, though this works well in my manual testing.

**Comment by bdarnell** [comment]
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 425 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRXEaJf3gy7jc-einkp:-KRXEaJgmI49mpGgiMBM:bc36kdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf421b2ea840ba0deb2af536b9c2a658dc64608/storage/replica.go#L425)):*
> ```Go
> 			// ready processing if the range only has a single replica (see
> 			// splitTriggerPostCommit). Doing so implies we need to be campaigning
> 			// that right hand side range when raft ready processing is performed.
> ```

I can't follow this comment. `splitTriggerPostCommit` already has a lot of special casing for when to campaign on the RHS, can add something there to campaign when `len(Replicas)==1` if we're not already in `StateLeader` by the time we process the split trigger?

---

*[storage/replica.go, line 1624 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRXFG3xux4y_Rj9jpQ7:-KRXFG3ySXulNCcCt-hh:b9d2em2) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf421b2ea840ba0deb2af536b9c2a658dc64608/storage/replica.go#L1624)):*
> ```Go
> 	// creation if we gossiped our store descriptor more than the election
> 	// timeout in the past.
> 	shouldCampaign := (r.mu.internalRaftGroup == nil) && r.store.canCampaignIdleReplica()
> ```

I think these checks should be done in `withRaftGroup` and `defaultProposeRaftCommandLocked` should simply pass `true` for `shouldCampaign`. The caller should decide whether it wants to expedite leader elections, and withRaftGroup should decide when that translates to a call to Campaign. 

---

*[storage/store.go, line 1142 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRRIs0pHfu-Lq5i2CKe:-KRXIi8wYndq1CoD4Qdq:b-c6e893) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bf497ce9e2b339be4cddb34f7b3ae08b153f28/storage/store.go#L1142)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `Store.GossipStore` is called after both `Node.connectGossip` and the gossiping of the node descriptor (which usually happens via `Node.validateStores -> Node.initNodeID -> Gossip.SetNodeDescriptor`). I could see about moving computing the timeout to earlier, but performing it here is both conservative and somewhat convenient.
</details>
OK. I misread the code here (mistakenly assuming that the store gossip is driven by a goroutine started in Store.Start, instead of coming from server.Node). I'd still feel better if the check you added that gossip were connected was here in Store.GossipStore instead of relying on the unclear chain between store and server. 

---

*[storage/store.go, line 1163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRRKjEUdvIHNVGaKvBX:-KRXGFcwr1gyvBllt8zo:b-dc3scy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7bf497ce9e2b339be4cddb34f7b3ae08b153f28/storage/store.go#L1163)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Should I make the timeout `2*RaftElectionTimeoutTicks`? That would mean there is a 6 second window at startup where we'll have to wait for campaigning to occur, but that seems ok. Or leave as is and add a comment?
</details>
Leave it as-is. If we're not going to randomize it I don't think it helps much to just make it a larger constant. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276)*

**Comment by petermattis** [comment]
Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 425 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRXEaJf3gy7jc-einkp:-KRZ3eKR0DUgh9Harudb:b-cfawf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf421b2ea840ba0deb2af536b9c2a658dc64608/storage/replica.go#L425)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I can't follow this comment. `splitTriggerPostCommit` already has a lot of special casing for when to campaign on the RHS, can add something there to campaign when `len(Replicas)==1` if we're not already in `StateLeader` by the time we process the split trigger?
</details>
I've expanded the comment. I think this could be cleaned up, but I'd prefer not to do so in this PR.

---

*[storage/replica.go, line 1624 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRXFG3xux4y_Rj9jpQ7:-KRZ2kA_3ppOdwllwF1f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf421b2ea840ba0deb2af536b9c2a658dc64608/storage/replica.go#L1624)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I think these checks should be done in `withRaftGroup` and `defaultProposeRaftCommandLocked` should simply pass `true` for `shouldCampaign`. The caller should decide whether it wants to expedite leader elections, and withRaftGroup should decide when that translates to a call to Campaign. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276)*

**Comment by tschottdorf** [comment]
Reviewed 4 of 4 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 425 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRXEaJf3gy7jc-einkp:-KRZ3G065ooCOmYQzAJL:bsc82fh) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf421b2ea840ba0deb2af536b9c2a658dc64608/storage/replica.go#L425)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I've expanded the comment. I think this could be cleaned up, but I'd prefer not to do so in this PR.
</details>
There's also a comment there that indicates that we're doing this only for some tests. Perhaps these tests just need a change (touching the RHS proactively)? Doesn't seem to be strong enough a reason to not do what you really want to do. That said, Ben seems to prefer to have this here, though I'm not clear why (having it in the proposal path seems natural to me). Doesn't have to be this PR, but file an issue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276)*

**Comment by petermattis** [comment]
Still waiting for an LGTM.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-:-KRZBonlcJisSJJ-oKLA:b82kj9q)*

**Comment by tschottdorf** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 1149 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRZEcNnPx-OqfqOIumH:-KRZEcNnPx-OqfqOIumI:bmjlhzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b947f303b3bd01d2391b103012aea0d22ad9a6e3/storage/store.go#L1149)):*
> ```Go
> 	// normal Raft election timeout.
> 	s.idleReplicaElectionTime.Lock()
> 	if s.idleReplicaElectionTime.at == (time.Time{}) {
> ```

There's a bit of nonlocality here - as far as I can tell, `GossipStore` is called for the first time eagerly in `(*Node).bootstrapStores` and then we also require here that gossip is connected, which is also (probably) due to higher level stuff. Perhaps that could be pointed out in a comment here (or even explicitly checked that Gossip is connected, just to be explicit). I'm fine with any way you find to make the picture more graspable when looking at this part of the code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-:-KRZFM2yVzg_vA9t1ZFB:bnfp4nl)*

**Comment by petermattis** [comment]
Review status: 3 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 425 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRXEaJf3gy7jc-einkp:-KRZu1GemAMM7cOBAPDh:bkd7lhu) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf421b2ea840ba0deb2af536b9c2a658dc64608/storage/replica.go#L425)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> There's also a comment there that indicates that we're doing this only for some tests. Perhaps these tests just need a change (touching the RHS proactively)? Doesn't seem to be strong enough a reason to not do what you really want to do. That said, Ben seems to prefer to have this here, though I'm not clear why (having it in the proposal path seems natural to me). Doesn't have to be this PR, but file an issue.
</details>
Issue filed.

---

*[storage/store.go, line 1149 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRZEcNnPx-OqfqOIumH:-KRZuOgnZbbe7_2PgRqO:b-k6s9tv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b947f303b3bd01d2391b103012aea0d22ad9a6e3/storage/store.go#L1149)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> There's a bit of nonlocality here - as far as I can tell, `GossipStore` is called for the first time eagerly in `(*Node).bootstrapStores` and then we also require here that gossip is connected, which is also (probably) due to higher level stuff. Perhaps that could be pointed out in a comment here (or even explicitly checked that Gossip is connected, just to be explicit). I'm fine with any way you find to make the picture more graspable when looking at this part of the code.
</details>
I've added a check that we're connected to gossip here and a comment explaining the initialization sequence. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276)*

**Comment by tschottdorf** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276)*

**Comment by bdarnell** [comment]
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 425 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRXEaJf3gy7jc-einkp:-KRap0pL7E6KrAoz8QOr:b-mk9all) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf421b2ea840ba0deb2af536b9c2a658dc64608/storage/replica.go#L425)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Issue filed.
</details>
I don't actually feel strongly about having this logic in withRaftGroup. I was originally agreeing with your TODO about making that move (and then later, I asked for *all* the campaigning logic to be synchronized here when it was split between the two). I originally thought we could avoid the boolean arg by doing something like `SoftState.Lead==0`, but that doesn't actually work. With the boolean arg, I don't think it's any better to do this here than in propose. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276)*

**Comment by petermattis** [comment]
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 425 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276#-KRXEaJf3gy7jc-einkp:-KRciTtrC9dn8lENmXRi:b-onz751) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf421b2ea840ba0deb2af536b9c2a658dc64608/storage/replica.go#L425)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I don't actually feel strongly about having this logic in withRaftGroup. I was originally agreeing with your TODO about making that move (and then later, I asked for *all* the campaigning logic to be synchronized here when it was split between the two). I originally thought we could avoid the boolean arg by doing something like `SoftState.Lead==0`, but that doesn't actually work. With the boolean arg, I don't think it's any better to do this here than in propose. 
</details>
Yeah, this isn't a huge ugly, but it feels a bit fragile. The interaction between the enqueuing of the replica for raft processing in `splitTriggerPostCommit` and the fact that it will get campaigned here is subtle.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9276)*
